src/daft-plan/src/physical_plan.rs

```diff
@@ -0,0 +1,133 @@
+#[cfg(feature = "python")]
+use {
+    crate::source_info::PyFileFormatConfig,
+    daft_core::python::schema::PySchema,
+    daft_dsl::python::PyExpr,
+    daft_table::python::PyTable,
+    pyo3::{pyclass, pymethods, PyAny, PyObject, PyRef, PyRefMut, PyResult, Python, ToPyObject},
+};
+
+use crate::physical_ops::*;
+
+#[derive(Debug)]
+pub enum PhysicalPlan {
+    TabularScanParquet(TabularScanParquet),
+    Filter(Filter),
+    Limit(Limit),
+}
+
+#[cfg_attr(feature = "python", pyclass)]
+struct PartitionIterator {
+    parts: Vec<PyObject>,
+    index: usize,
+}
+
+#[cfg(feature = "python")]
+#[pymethods]
+impl PartitionIterator {
+    fn __iter__(slf: PyRef<'_, Self>) -> PyRef<'_, Self> {
+        slf
+    }
+    fn __next__(mut slf: PyRefMut<'_, Self>) -> Option<PyObject> {
+        let index = slf.index;
+        slf.index += 1;
+        slf.parts.get(index).map(|part| part.clone_ref(slf.py()))
+    }
+}
+
+#[cfg(feature = "python")]
+impl PhysicalPlan {
```

Comment:
>thoughts on implementing this part in python instead for brevity?

I briefly mentioned this in the PR description, but there are a few motivators for keeping at least the physical plan traversal in Rust, and maybe the calling of the `physical_plan.py` functions in Rust as well:
1. We can gradually port the scheduling and execution logic for individual operators over from Python to Rust; I'm loosely thinking that we could create all-Rust-backed `Instruction` implementations + `PartitionTask`/`Runner` logic to execute a pipeline of such instructions without the GIL.
2. By driving the Python-based physical plan scheduling from the Rust side, we avoid having to expose the Rust physical operators (or Python-oriented views of them) to Python, which would otherwise be a decent bit of pyo3 work.

(2) is the main driver since (1) is a bit hand-wavy at this point, but maybe there's a good middle ground here with a Python shim layer that's oriented to the Rust physical plan? We could then do away with a good bit of these pyo3 function calls to set up the arguments for the `physical_plan.*()` calls. For example, if we had something like

```python
def filter(input: InProgressPhysicalPlan[PartitionT], predicate: PyExpr) -> InProgressPhysicalPlan[PartitionT]:
    expr = Expr._from_pyexpr(predicate)
    expr_projection = ExpressionsProjection([expr])
    return physical_plan.pipeline_instruction(input, Filter(expr_projection), ResourceRequest())
```

then the Rust-side would be reduced to:

```rust
            PhysicalPlan::Filter(Filter { input, predicate }) => {
                let upstream_iter = input.to_partition_tasks(py)?;
                let py_iter = py
                    .import(pyo3::intern!(py, "daft.execution.rust_physical_plan"))?
                    .getattr(pyo3::intern!(py, "filter"))?
                    .call1((upstream_iter, PyExpr::from(predicate.clone())))?;
                Ok(py_iter.into())
            }
```

Reply:
We can also shim as needed per op, so if a shim would be nicer for the `Agg` op, definitely feel free to add it!